codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -445,8 +506,9 @@ public FlowControlledBase(final ChannelHandlerContext ctx, final Http2Stream str
             this.endOfStream = endOfStream;
             this.stream = stream;
             this.promise = promise;
-            // Ensure error() gets called in case something goes wrong after the frame is passed to Netty.
-            promise.addListener(this);
+            if (!promise.isVoid()) {
```

Comment:
The base class really only needs the promise in the `writeComplete` method.  Maybe instead of the base class owning the promise, you define an abstract method `promise()`. Headers will listen to the promise in its constructor, so it will always have to ensure that the promise is non-Void. Data will lazy create it as-needed. Maybe the base class could define:
```java
ChannelPromise toListenerablePromise(ChannelPromise promise) {
    return promise.isVoid() ? ctx.newPromise() : promise;
}
```